xs/src/libslic3r/PrintGCode.cpp

```diff
@@ -492,118 +492,137 @@ PrintGCode::process_layer(size_t idx, const Layer* layer, const Points& copies)
+                auto extruder_id = region->config.perimeter_extruder-1;
+                // Casting away const just to avoid double dereferences
+                for(auto* perimeter_coll : const_cast<LayerRegion*>(layerm)->perimeters) {
+                    if(perimeter_coll->length() == 0) continue;  // this shouldn't happen but first_point() would fail
                     
-                    # init by_extruder item only if we actually use the extruder
-                    $by_extruder{$extruder_id} //= [];
-                    
-                    # $perimeter_coll is an ExtrusionPath::Collection object representing a single slice
-                    for my $i (0 .. $n_slices) {
-                        if (
-                            # $perimeter_coll->first_point does not fit inside any slice
-                            $i == $n_slices 
-                            # $perimeter_coll->first_point fits inside ith slice
-                            || $point_inside_surface->($i, $perimeter_coll->first_point)) {
-                            $by_extruder{$extruder_id}[$i] //= { perimeters => {} };
-                            $by_extruder{$extruder_id}[$i]{perimeters}{$region_id} //= [];
-                            push @{ $by_extruder{$extruder_id}[$i]{perimeters}{$region_id} }, @$perimeter_coll;
-                            last;
+                    // perimeter_coll is an ExtrusionPath::Collection object representing a single slice
+                    for(auto i = 0U; i < n_slices; i++){
+                        if (// perimeter_coll->first_point does not fit inside any slice
+                            i == n_slices - 1
+                            // perimeter_coll->first_point fits inside ith slice
+                            || point_inside_surface(i, perimeter_coll->first_point())) {
+                            std::get<0>(by_extruder[extruder_id][i])[region_id].append(*perimeter_coll);
+                            break;
                         }
                     }
                 }
             }
             
-            # process infill
-            # $layerm->fills is a collection of ExtrusionPath::Collection objects, each one containing
-            # the ExtrusionPath objects of a certain infill "group" (also called "surface"
-            # throughout the code). We can redefine the order of such Collections but we have to 
-            # do each one completely at once.
-            foreach my $fill (@{$layerm->fills}) {
-                next if $fill->empty;  # this shouldn't happen but first_point() would fail
-                
-                # init by_extruder item only if we actually use the extruder
-                my $extruder_id = $fill->[0]->is_solid_infill
-                    ? $region->config->solid_infill_extruder-1
-                    : $region->config->infill_extruder-1;
+            // process infill
```

Comment:
I wonder if it would be better to break this (and the perimeters) into their own methods.